torch_xla/csrc/runtime/pjrt_computation_client.h

```diff
@@ -37,17 +37,17 @@ class PjRtComputationClient : public ComputationClient {
   std::optional<xla::OpSharding> GetDataSharding(DataPtr handle) override;
 
   std::vector<DataPtr> TransferToServer(
-      absl::Span<const TensorSource> tensors) override;
+      absl::Span<const std::shared_ptr<const TensorSource>> tensors) override;
```

Comment:
Should we require the caller to manage the ownership & lifetime of TensorSource*, i.e., `const TensorSource*` instead, or it's necessary to ensure that the memory is held during the client ops and in the client?

Reply:
Good question. PJRT lets us tie the lifetime of an object to an operation by capturing it in a callback. `std::function`s have to be copyable, so `shared_ptr` is our best choice here. `TensorSource` itself may be expensive or impossible to copy.

The caller of `TransferToServer` will be much shorter-lived than the actual transfer, so ownership should pass down. We could tighten up the interface here and consume a `unique_ptr` since we only need copyability within the implementation of `TransferToServer`. What do you think?